collector/app.go

```diff
@@ -1,18 +1,91 @@
 package collector
 
 import (
+	"strings"
+
 	norman "github.com/rancher/norman/types"
-	client "github.com/rancher/types/client/project/v3"
+	rancher "github.com/rancher/types/client/project/v3"
 	log "github.com/sirupsen/logrus"
 )
 
-func GetAppsCollection(c *CollectorOpts, url string) *client.AppCollection {
+const (
+	AppTemplateExternalIDPrefix = "catalog://?"
+	AppCatalogLibrary           = "library"
+	AppCatalogSystemLibrary     = "system-library"
+)
+
+var (
+	AppRancherCatalogs = []string{AppCatalogLibrary, AppCatalogSystemLibrary}
+)
+
+type AppTemplate struct {
+	State string                 `json:"state"`
+	Apps  map[string]*LabelCount `json:"apps"`
+}
+
+type App struct {
+	Total    int                     `json:"total"`
+	Active   int                     `json:"active"`
+	Catalogs map[string]*AppTemplate `json:"rancheCatalogs"`
+}
+
+func (a App) RecordKey() string {
+	return "app"
+}
+
+func (a App) Collect(c *CollectorOpts) interface{} {
+	opts := NonRemoved()
+	opts.Filters["all"] = "true"
+
+	a.Catalogs = map[string]*AppTemplate{}
+
+	for _, catalog := range AppRancherCatalogs {
+		a.Catalogs[catalog] = &AppTemplate{
+			State: GetAppCatalogState(c, catalog),
+			Apps:  map[string]*LabelCount{},
+		}
+	}
+
+	log.Debug("Collecting Projects")
+	projectList, err := c.Client.Project.List(&opts)
+
+	if err != nil {
+		log.Errorf("Failed to get Projects err=%s", err)
+		return nil
+	}
+
+	for _, project := range projectList.Data {
+		log.Debug("Collecting Apps")
+		appsData := GetAllAppData(c, project.Links["apps"])
+		if appsData != nil {
+			for _, app := range appsData {
+				externalID := SplitAppExternalID(app.ExternalID)
+				if a.Catalogs[externalID["catalog"]] == nil {
```

Comment:
Is `externalID[catalog]` guaranteed to have values ? Same with template and version below.

Reply:
Updated